### PR TITLE
caffe2: init at 0.8.1

### PR DESCRIPTION
**Pull request by yuriaisaka**
###### Motivation for this change

Caffe2 is a lightweight, modular, and scalable deep learning framework developed at Facebook.
It is a successor to the widely used Caffe framework, and provides both C++ and Python APIs.

Requests for review:

(1) package location

I've tentatively put the package to `pkgs/development/library/science/math`,
although it might fit nicer to `pkgs/development/library/science/machine-learning`
if there's such a directory.

Notes:
- There's a directory named `pkgs/applications/science/machine-learning/`
- `caffe` resides in: `pkgs/applications/science/math/`
- Other major deep learning frameworks (notably `tensorflow` & `pytorch`) are registered as python packages.

(2) CUDA issues

- Caffe2 v0.8.1 (the version packaged in this PR) does not officially support Cuda Toolkit9
- Cuda Toolkit8 does not officially support gcc6
- Caffe2/CPU v0.8.1 compiles and seems to work fine both with gcc5 and gcc6.
- I haven't been able to compile Caffe2/GPU v0.8.1 with Cuda Toolkit8 and gcc5.5 from `18.03pre` (`overrideCC stdenv pkgs.gcc5`). It compiles (and seems to work fine) with Cuda Toolkit8 and gcc5.4 from `release-17.09`.
- The errors from Caffe2/GPU v0.8.1 with Cuda Toolkit8 and gcc5.5 seems to be arising from AVX intrinsics, e.g.:
  ```
  Building NVCC (Device) object caffe2/CMakeFiles/Caffe2_GPU.dir/utils/Caffe2_GPU_generated_math_gpu.cu.o
  /nix/store/0l7rjsr59d7l8pr0gxvf37jx8gnva5fi-gcc-5.5.0/lib/gcc/x86_64-unknown-linux-gnu/5.5.0/include/avx512fintrin.h(9220): error: argument of type "const void *" is incompatible with parameter of type "const float *"
  ``` 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/libraries/science/math/caffe2/default.nix [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+      sha256 = "05gi58dirvc8fgm0avpydvidzsbh2zrzgfaq671ym09f6dz0bcgz";
+    };
+    dst = "pybind11";
+  };
+in
+
+stdenv.mkDerivation rec {
+  name = "caffe2-${version}";
+  version = "0.8.1";
+  src = fetchurl {
```
Comment (joachifm) [inline pkgs/development/libraries/science/math/caffe2/default.nix]: Consider re-using `version` when fetching the source, for easier bumping. Also consider
using the `fetchFromGitHub` helper, it is a little bit more succinct & "safer" when using autogenerated archives.

Comment (yuriaisaka) [reply pkgs/development/libraries/science/math/caffe2/default.nix]: Makes sense, thanks! Have updated the PR.

**Comment by joachifm** [comment]
@GrahamcOfBorg build caffe2

**Comment by yuriaisaka** [comment]
600b1d3 tries to fix the build failure on x86_64-darwin.
Since I don't have a working Darwin environment, shall remove darwin support for now if this attempt fails.

As for the build failure on aarch64-linux, probably shall `pkgs/development/libraries/opencv/3.x.nix` check if `stdenv.system` is one of (`x86_64-linux`, `i686-linux`, `x86_64-darwin`) before calling `installExtraFiles ippicv` at `3.x.nix:177`? As far as I understand, opencv3 works without IntelCV.

File: pkgs/development/libraries/science/math/caffe2/update_clang_cvtsh_bugfix.patch [hunk 2]
```diff
@@ -0,0 +1,55 @@
+diff --git a/caffe2/perfkernels/cvtsh_ss_bugfix.h b/caffe2/perfkernels/cvtsh_ss_bugfix.h
```
Comment (joachifm) [inline pkgs/development/libraries/science/math/caffe2/update_clang_cvtsh_bugfix.patch]: Is there an external source for this patch? It is generally preferred to fetch patches (using e.g., `fetchpatch`).

File: pkgs/development/libraries/science/math/caffe2/default.nix [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    ++ lib.optional useOpenmp openmp
+    ++ lib.optional useOpencv3 opencv3
+    ++ lib.optional useLeveldb leveldb
+    ++ lib.optional useLmdb lmdb
+    ++ lib.optional useRocksdb rocksdb
+    ++ lib.optional useZeromq zeromq
+  ;
+  propagatedBuildInputs = [ numpy future six python-protobuf pydot ];
+
+  patches = if stdenv.cc.isClang then [ ./update_clang_cvtsh_bugfix.patch ] else [ ];
```
Comment (joachifm) [inline pkgs/development/libraries/science/math/caffe2/default.nix]: The pattern `if foo then [ x ] else [ ]` can be reduced to `optional foo x`

File: pkgs/development/libraries/science/math/caffe2/update_clang_cvtsh_bugfix.patch (hunk 2 above)
Comment (yuriaisaka) [reply pkgs/development/libraries/science/math/caffe2/update_clang_cvtsh_bugfix.patch]: Thanks for the suggestion. As a single patch, I suspect not.

557f877 in https://github.com/caffe2/caffe2/commits/master/caffe2/perfkernels/cvtsh_ss_bugfix.h is in v0.8.1 and there're 4-5 patches upstream to cover various versions/flavors of clang.

But then, what the patch does is to replace `cvtsh_ss_bugfix.h` at v0.8.1 by that of the current master, so I could fetch the latest `cvtsh_ss_bugfix.h` and override by it in one of the build hooks.

File: pkgs/development/libraries/science/math/caffe2/default.nix (hunk 3 above)
Comment (yuriaisaka) [reply pkgs/development/libraries/science/math/caffe2/default.nix]: I should have noticed that, thanks.

**Pull request merged**